sycl/include/CL/sycl/queue.hpp

```diff
@@ -355,36 +355,71 @@ class __SYCL_EXPORT queue {
     return submit([=](handler &CGH) { CGH.prefetch(Ptr, Count); });
   }
 
+  // having _TWO_ mid-param #ifdefs makes the functions very difficult to read.
+  // Here we simplify the &CodeLoc declaration to be CODELOCPARAM(&CodeLoc) and
+  // CODELOCARG(&CodeLoc) Similarly, the KernelFunc param is simplified to be
+  // KERNELFUNCPARAM(KernelFunc) Once the queue kernel functions are defined,
+  // these macros are #undef immediately. Compare this first definition of
+  // single_task here, with the second one below.
+
+  // replace CODELOCPARAM(&CodeLoc) with nothing
+  // or :   , const detail::code_location &CodeLoc =
+  // detail::code_location::current()
+  // replace CODELOCARG(&CodeLoc) with nothing
+  // or :  const detail::code_location &CodeLoc = {}
+
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+#define CODELOCPARAM(a)                                                        \
+  , const detail::code_location a = detail::code_location::current()
+
+#define CODELOCARG(a)
+#else
+#define CODELOCPARAM(a)
+
+#define CODELOCARG(a) const detail::code_location a = {}
+#endif
+// replace KERNELFUNCPARAM(KernelFunc) with   KernelType KernelFunc
+//                                     or     const KernelType &KernelFunc
+#ifdef __SYCL_NONCONST_FUNCTOR__
+#define KERNELFUNCPARAM(a) KernelType a
+#else
+#define KERNELFUNCPARAM(a) const KernelType &a
+#endif
+
   /// single_task version with a kernel represented as a lambda.
   ///
   /// \param KernelFunc is the Kernel functor or lambda
   /// \param CodeLoc contains the code location of user code
   template <typename KernelName = detail::auto_name, typename KernelType>
-  event single_task(
-      KernelType KernelFunc
-#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
-      ,
-      const detail::code_location &CodeLoc = detail::code_location::current()
-#endif
-  ) {
-#ifdef DISABLE_SYCL_INSTRUMENTATION_METADATA
-    const detail::code_location &CodeLoc = {};
-#endif
+  event single_task(KERNELFUNCPARAM(KernelFunc) CODELOCPARAM(&CodeLoc)) {
```

Comment:
This declaration of single_task is much easier to read than the next one, by virtue of these macros. @bader , @romanovvlad let me know your thoughts. I can remove them, or use them in the other affected declarations in queue.hpp.

Reply:
Do not have any preference on this.

Reply:
Then I will tidy them up with the macros in these declarations in queue.hpp.    There is also a in-param-list #ifdef in some of the declarations in handler.hpp but they only have one #ifdef set, not two.   So I'm not going to adjust those.